### ISSUE TITLE
Checking of @EnsuresNonNull / @RequiresNonNull varies with use of fluent API

### ISSUE THREAD

**Issue by msridhar** [typetools/checker-framework#2465]
I am getting some unexpected results using `@EnsuresNonNull` / `@RequiresNonNull` with a fluent API.  This is with Checker Framework 2.8.0.  Consider the following example:
```java
import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
import org.checkerframework.checker.nullness.qual.RequiresNonNull;
public class Library {
    static class Example {
        @MonotonicNonNull String s;
        public Example() {}
        @EnsuresNonNull("this.s")
        public Example setS(String s1) {
            this.s = s1;
            return this;
        }
        @RequiresNonNull("this.s")
        public void print() {
            System.out.println(this.s.toString());
        }
    }
    static void withVars() {
        Example example = new Example();
        example.setS("hi");
        example.print();
    }
    static void fluent1() {
        Example example = new Example();
        example
            .setS("hi")
            .print();
    }
    static void fluent2() {
        (new Example())
            .setS("hi")
            .print();
    }
}
```
I would expect this code to typecheck with no errors.  However, when running Nullness Checker I get the following error inside `fluent1()`:
```
Library.java:43: error: [contracts.precondition.not.satisfied] example.setS("hi").print's precondition about 'example.setS("hi").s' is not satisfied
            .print();
                  ^
1 error
```
Adding an `@Pure` annotation to `setS()` does not fix the problem.  Also, it is a bit surprising that `fluent1()` does not type check, but getting rid of the `example` local variable in `fluent2()` makes the code type check.

**Comment by smillst** [comment]
The precondition `@EnsuresNonNull("this.s")` says nothing about the nullness of the returned object's field `s`.  So, I would expect both `fluent1` and `fluent2` to issue errors.  
I'm not sure why `fluent2` type checks.   The following additional method also does not issue an error:
```
    static void fluent3() {
        new Example()
            .setS("hi");
        new Example()
            .print();
    }
```
So, I think there's some sort of bug with constructors and pre/postconditions.

 If you change the precondition to ` @EnsuresNonNull({"this.s", "this.setS(#1).s"})` and add `@Pure`, then `fluent1` type checks, but the following error is issued:
```
Library.java:11: error: [contracts.postcondition.not.satisfied] the postcondition EnsuresNonNull about 'this.setS(s1).s' of this method is not satisfied
        public Example setS(String s1) {
                       ^
1 error
```
It would be nice if we had some sort of flow expression syntax that refers to the return object, so that this postcondition could be verified.

**Comment by wmdietl** [comment]
One thing that the specification of `setS` doesn't say is that the returned `Example` object is the same as `this`.
The post-condition only says that `this.s` will be non-null, it doesn't say anything about what object it returns.

In `withVars` we substitute `example` for `this` and everything is ok.

In neither of the fluid uses do we actually know that the object returned from `setS` has the field set.
So isn't it surprising that `fluent2` doesn't raise an error?

I would try adding a post-condition to `setS` that expresses that the returned object has a non-null field:
@EnsuresNonNull("<ret>.s").
However, I don't think we have the vocabulary for that.

More generally for fluent APIs, we might want an annotation that expresses that the returned object is the same as the receiver of the call.

**Comment by msridhar** [comment]
Thanks for the quick feedback @smillst and @wmdietl!  So it looks like there are two separate issues here.  One is that there is a bug with constructors and pre/post-conditions.  The other is that the Nullness Checker currently doesn't have a way to write `@EnsuresNonNull("return.s")`, so we can't express the right post-condition for `setS()`.  Should I open another issue then?  Maybe we can keep this one about the constructor issue and have a separate issue about postconditions on return values?

> More generally for fluent APIs, we might want an annotation that expresses that the returned object is the same as the receiver of the call.

Yes, I could see that being helpful in some cases.  But probably if some way gets added to talk about the return value in `@EnsuresNonNull` that would be good enough for most cases.

Let me also say at the moment I'm not blocked on either of these issues.

**Comment by msridhar** [comment]
Nothing urgent, but @smillst @wmdietl let me know if you want me to refocus this issue on the unsoundness, which is likely a more urgent issue than support for fluent APIs.

**Comment by wmdietl** [comment]
I would suggest opening a new issue that just focusses on the unsoundness you identified.
This issue can contain the longer discussion relevant for the fluent API improvements.

**Comment by msridhar** [comment]
Sure thing, opened #2470
